File: JobTracker/Services/ActivityLog.swift
//
//  ActivityLog.swift
//  JobTracker
//
//  In-app activity log so the user can watch what ingestion and sync are
//  doing (parse results, per-email classification, skips, errors). Entries
//  are also mirrored to the unified logging system for Console.app.
//

import Foundation
import os

@MainActor
@Observable
final class ActivityLog {
    static let shared = ActivityLog()

    enum Level: String {
        case info = "•"
        case success = "✓"
        case warning = "△"
        case error = "✕"
    }

    struct Entry: Identifiable {
        let id = UUID()
        let date: Date
        let level: Level
        let message: String
    }

    private(set) var entries: [Entry] = []
    private static let cap = 5000
    private let logger = Logger(subsystem: "com.jobtracker", category: "activity")

    private init() {}

    func info(_ message: String) { append(.info, message) }
    func success(_ message: String) { append(.success, message) }
    func warning(_ message: String) { append(.warning, message) }
    func error(_ message: String) { append(.error, message) }

    private func append(_ level: Level, _ message: String) {
        entries.append(Entry(date: Date(), level: level, message: message))
        if entries.count > Self.cap {
            entries.removeFirst(entries.count - Self.cap)
        }
        switch level {
        case .error:   logger.error("\(message, privacy: .public)")
        case .warning: logger.warning("\(message, privacy: .public)")
        default:       logger.info("\(message, privacy: .public)")
        }
    }

    func clear() { entries.removeAll() }

    /// Plain-text dump for the copy button.
    var text: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm:ss"
        return entries.map { "[\(formatter.string(from: $0.date))] \($0.level.rawValue) \($0.message)" }
            .joined(separator: "\n")
    }
}

File: JobTracker/Services/ActivityLog.cs
// In-app activity log so the user can watch what ingestion/sync are doing
// (parse results, per-email classification, skips, errors). Entries are
// also mirrored to System.Diagnostics.Trace.
//
// Framework-agnostic by design (no WPF Dispatcher dependency) so it's usable
// from background sync tasks and from unit tests alike; subscribe to
// Changed and marshal to the UI thread from the ViewModel side.

using System.Diagnostics;
using System.Threading;

namespace JobTracker.Services;

public sealed class ActivityLog
{
    public static readonly ActivityLog Shared = new();

    public enum Level { Info, Success, Warning, Error }

    public sealed record Entry(Guid Id, DateTimeOffset Date, Level Level, string Message);

    private const int Cap = 5000;
    private readonly Lock _lock = new();
    private readonly List<Entry> _entries = [];

    /// Raised (possibly off the UI thread) whenever entries change.
    public event Action? Changed;

    private ActivityLog() { }

    public IReadOnlyList<Entry> Entries
    {
        get { lock (_lock) { return _entries.ToList(); } }
    }

    public void Info(string message) => Append(Level.Info, message);
    public void Success(string message) => Append(Level.Success, message);
    public void Warning(string message) => Append(Level.Warning, message);
    public void Error(string message) => Append(Level.Error, message);

    private void Append(Level level, string message)
    {
        lock (_lock)
        {
            _entries.Add(new Entry(Guid.NewGuid(), DateTimeOffset.Now, level, message));
            if (_entries.Count > Cap)
            {
                _entries.RemoveRange(0, _entries.Count - Cap);
            }
        }
        switch (level)
        {
            case Level.Error: Trace.TraceError(message); break;
            case Level.Warning: Trace.TraceWarning(message); break;
            default: Trace.TraceInformation(message); break;
        }
        Changed?.Invoke();
    }

    public void Clear()
    {
        lock (_lock) { _entries.Clear(); }
        Changed?.Invoke();
    }

    /// Plain-text dump for the copy button.
    public string Text
    {
        get
        {
            lock (_lock)
            {
                return string.Join('\n', _entries.Select(e =>
                    $"[{e.Date:HH:mm:ss}] {e.Level.Glyph()} {e.Message}"));
            }
        }
    }
}

public static class ActivityLogLevelExtensions
{
    public static string Glyph(this ActivityLog.Level level) => level switch
    {
        ActivityLog.Level.Info => "•",
        ActivityLog.Level.Success => "✓",
        ActivityLog.Level.Warning => "△",
        ActivityLog.Level.Error => "✕",
        _ => "•",
    };
}
